### PR TITLE
AOA multiplexing for Android Proxy

### PR DESCRIPTION
**Pull request by shiniwat**
This proposal aims to add AOA multiplexing capability to SdlRouterService.

### CODE REVIEW

File: proposals/0095-AOA-multiplexing.md [hunk 1]
```diff
@@ -14,161 +14,154 @@ This proposal aims to add AOA multiplexing capability to SdlRouterService.
 
 Currently SdlRouterService works for MultiplexTransport, which supports Bluetooth transport as the underlying transport. This proposal extends SdlRouterService to support both Bluetooth and AOA (Android Open Accessory) transports, so that multiple SDL applications can share the AOA transport without having to worry about USB connection.
 AOA protocol basically allows only a single application to access the USB accessory -- if multiple applications are designed to use the same accessory, the Android system asks which application should use the accessory and shows a permission dialog. This proposal allows multiple SDL applications to share the USB accessory.
+In addition, this proposal addresses extension of SdlRouterService to support simultaneous multiple transports, and AOA multiplexing will work as a part of simultaneous multiple transports. This proposal mainly assumes the scenario where utilizing Bluetooth multiplexing transport for RPC service, and utilizing AOA multiplexing transport for audio/video services.
+As you see the ![0141-multiple-transports proposal](0141-multiple-transports.md), the secondary transport will be specified by SDLCore, however.
```
Comment (theresalech) [inline proposals/0095-AOA-multiplexing.md]: Please update to "As you see in the [0141-multiple-transports proposal](0141-multiple-transports.md),"

File: proposals/0095-AOA-multiplexing.md [hunk 2]
```diff
@@ -14,161 +14,154 @@ This proposal aims to add AOA multiplexing capability to SdlRouterService.
...
-- The affected component is Android proxy only. No change is required for SDL Core.
+- MultiplexTransportConfig is extended to have "highBandwidthRequired" flag, which should be turned on for audio/video apps.
+- If SDL Core supports secondary transport, proxy will be notified with onEnableSecondaryTransport. If the secondary transport is AOA_USB, the proxy takes it as AOA multiplexing.
+- If AOA_USB is the expected secondary transport, SdlProxy puts off starting audio/video session until AOA transport becomes available.
+- Make sure the AOA multiplexing can be used with older SDLCore, which does not support simultaneous multiple transports. When running with older SDLCore and the app specified "highBandwidthRequired" flag, MultiplexTransport will use AOA transport for all service types. This means the app cannot be registered until AOA transport is available.
+- AOA multiplexing will be solely handled by Android Proxy. From SDLCore's perspective, AOA multiplexing will be handled as AOA_USB transport type, and the existing AOA transport adapter on SDLCore should work for AOA multipelxing.
  
+The sequence diagram running on secondary transport are shown below:
```
Comment (theresalech) [inline proposals/0095-AOA-multiplexing.md]: Please update to "...is shown below:"

File: proposals/0095-AOA-multiplexing.md [hunk 3]
```diff
@@ -14,161 +14,154 @@ This proposal aims to add AOA multiplexing capability to SdlRouterService.
...
-    ((MultiplexTransportConfig) transport).setTransportType(TransportType.MULTIPLEX_AOA);
+    ((MultiplexTransportConfig) transport).setHighBandwidthRequired(true);
     proxy = new SdlProxyALM(this,
                     ....
                     transport);
 ```
 
-### SdlRouterService holds newly added Handler for AOA clients
-Extends SdlRouterService by adding AoaRouterHandler:
+### When handing StartSessionACK control frame, checks to see if secondary transport is supported.
```
Comment (theresalech) [inline proposals/0095-AOA-multiplexing.md]: Should this be updated to "When handling" ?

File: proposals/0095-AOA-multiplexing.md [hunk 4]
```diff
@@ -14,161 +14,154 @@ This proposal aims to add AOA multiplexing capability to SdlRouterService.
...
+            // corrsponding Binder will use aoaTransport
+        }
+    }
 }
 ```
 
 ### The app that has actual UsbAccessory's permission sends ParcelFileDescriptor to active SdlRouterService
-We discussed about a potential issue where multiple SDL applications have the same accessory_filter settings.
+When a user plugs in their device to an AOA connection, they are prompted with selection dialog. The dialog contains all apps that support the currently connected accessory. In this dialog there is also an option to always open a specific app for that accessory.
+We basically assume the app that receives AOA intent gets the accessory permission, and then the app works as the router for AOA transport. if the user picks an app to always receive the AOA intent and it propagates the router service, all multiplexed connections will go through that apps router service regardless if there is a newer router service present. If by chance that is not a trusted router service, no other apps will ever connect until the user clears the flag.
```
Comment (theresalech) [inline proposals/0095-AOA-multiplexing.md]: Please update the second sentence to "If the user... will go through that app's router service regardless of if there..."

File: proposals/0095-AOA-multiplexing.md [hunk 5]
```diff
@@ -14,161 +14,154 @@ This proposal aims to add AOA multiplexing capability to SdlRouterService.
...
 
 This feature also increases the IPC transaction between TransportBroker and SdlRouterService. While Android system has Binder Transaction Limit, which is explained at [TransactionTooLargeException Android document](https://developer.android.com/reference/android/os/TransactionTooLargeException.html), we won’t run into TransactionTooLargeException cases in real scenario based on our test, unless underlying transport has a fatal error. The fatal error case would be, for example, the case where we cannot write Bluetooth socket and/or USB's ParcelFileDescriptor for some reason. Those fatal error cases can be discussed outside of this proposal.
 
+Also note that this proposal assumes MultiplexBluetoothTransport is available prior to MultiplexAoaTransport becomes available. Otherwise we cannot check if the SDL Core supports simultaneous multiple transports. What if the head unit is configured without Bluetooth? What if user did not setup Bluetooth pairing beforehand? Such case may need to be discussed, but for now it is out of scope in this proposal.
```
Comment (theresalech) [inline proposals/0095-AOA-multiplexing.md]: Please update to "...prior to when MultiplexAoaTransport becomes available."

**Comment by shiniwat** [comment]
Hi @theresalech, thanks for the review! I have updated all points you suggested. I hope it is in the good shape now.

**Pull request merged**